--- a/drivers/gpu/drm/drm_fb_helper.c	2015-01-16 17:39:12.688025650 +0000
+++ b/drivers/gpu/drm/drm_fb_helper.c	2015-01-16 22:07:22.031761349 +0000
@@ -760,9 +760,9 @@
 		var->transp.offset = 15;
 		break;
 	case 16:
-		var->red.offset = 11;
+		var->blue.offset = 11;
 		var->green.offset = 5;
-		var->blue.offset = 0;
+		var->red.offset = 0;
 		var->red.length = 5;
 		var->green.length = 6;
 		var->blue.length = 5;
@@ -770,9 +770,9 @@
 		var->transp.offset = 0;
 		break;
 	case 24:
-		var->red.offset = 16;
+		var->blue.offset = 16;
 		var->green.offset = 8;
-		var->blue.offset = 0;
+		var->red.offset = 0;
 		var->red.length = 8;
 		var->green.length = 8;
 		var->blue.length = 8;
@@ -780,9 +780,9 @@
 		var->transp.offset = 0;
 		break;
 	case 32:
-		var->red.offset = 16;
+		var->blue.offset = 16;
 		var->green.offset = 8;
-		var->blue.offset = 0;
+		var->red.offset = 0;
 		var->red.length = 8;
 		var->green.length = 8;
 		var->blue.length = 8;
@@ -1073,9 +1073,9 @@
 		info->var.transp.length = 0;
 		break;
 	case 15:
-		info->var.red.offset = 10;
+		info->var.blue.offset = 10;
 		info->var.green.offset = 5;
-		info->var.blue.offset = 0;
+		info->var.red.offset = 0;
 		info->var.red.length = 5;
 		info->var.green.length = 5;
 		info->var.blue.length = 5;
@@ -1083,18 +1083,18 @@
 		info->var.transp.length = 1;
 		break;
 	case 16:
-		info->var.red.offset = 11;
+		info->var.blue.offset = 11;
 		info->var.green.offset = 5;
-		info->var.blue.offset = 0;
+		info->var.red.offset = 0;
 		info->var.red.length = 5;
 		info->var.green.length = 6;
 		info->var.blue.length = 5;
 		info->var.transp.offset = 0;
 		break;
 	case 24:
-		info->var.red.offset = 16;
+		info->var.blue.offset = 16;
 		info->var.green.offset = 8;
-		info->var.blue.offset = 0;
+		info->var.red.offset = 0;
 		info->var.red.length = 8;
 		info->var.green.length = 8;
 		info->var.blue.length = 8;
